src/vm/assembly.cpp

```diff
@@ -1965,6 +1965,10 @@ INT32 Assembly::ExecuteMainMethod(PTRARRAYREF *stringArgs, BOOL waitForOtherThre
     IfFailThrow(hr);
     
     END_ENTRYPOINT_THROWS;
+
+    // Sync the latched exit code with the return value of main function.
+    SetLatchedExitCode(iRetVal);
```

Comment:
I just ran the sample program and `echo $?` gave me 123. Note I've removed the check and my sync code before testing. So it works.